### ISSUE TITLE
Cannot authenticate Azure SQL DB due to InvalidAuthenticationTokenAudience error (aud: `https://database.windows.net/`) 

### ISSUE THREAD

**Issue by mir4ef** [Azure/ms-rest-nodeauth#81]
Hi,

we have a node app and we are trying to use Azure Active Directory Authentication for authentication with Azure SQL DB. we use [`sequelize`](https://sequelize.org/v5/class/lib/sequelize.js~Sequelize.html#instance-constructor-constructor) to establish the connection. under the hood, it uses [`tedious`](https://tediousjs.github.io/tedious/api-connection.html), which uses this package (`@azure/ms-rest-nodeauth`). Here are the versions we use:

- `nodejs` - v10.16.3
- `@azure/ms-rest-nodeauth` - v2.0.5
- `sequelize` - v5.18.4
- `tedious` - v6.4.0

The configuration we pass to `sequelize` is as follow:

```javascript
const Sequelize = require('sequelize');

const DB_CONFIG = {
    username: process.env.DB_USERNAME,
    password: process.env.DB_PASSWORD,
    name: process.env.DB_NAME,
    settings: {
        host: process.env.DB_HOST,
        port: process.env.DB_PORT,
        dialect: 'mssql',
        dialectOptions: {
            authentication: {
              type: 'azure-active-directory-password',
            },
            options: {
                encrypt: true,
            },
            requestTimeout: 600000,
        },
        pool: {
            max: 10,
            min: 0,
            idle: 5000,
            acquire: 5000,
        },
    },
};

const sequelize = new Sequelize(
    DB_CONFIG.name,
    DB_CONFIG.username,
    DB_CONFIG.password,
    DB_CONFIG.settings
);
```

however, we are facing an authentication problem coming from `@azure/ms-rest-nodeauth`. The error that gets thrown seems some sort of improper handling of the response from `https://management.azure.com/tenants?api-version=2016-06-01`:
```
TypeError: tenants.value is not iterable
```
from the [for loop](https://github.com/Azure/ms-rest-nodeauth/blob/adea8415de97a8076006fbd9948123d09dfdd78d/lib/subscriptionManagement/subscriptionUtils.ts#L94) inside `buildTenantList` method (inside `subscriptionUtils.ts`), which expects `value` to be always present and iterable. Nevertheless, the actual problem is the response from the above call - the audience doesn't match. The actual response body (`res.parsedBody`) is:

```json
{
  error: {
    code: 'InvalidAuthenticationTokenAudience',
    message: 'The access token has been obtained for wrong audience or resource \'https://database.windows.net/\'. It should exactly match with one of the allowed audiences \'https://management.core.windows.net/\',\'https://management.core.windows.net\',\'https://management.azure.com/\',\'https://management.azure.com\'.',
  },
}
``` 

The audience in the `accessToken` we get is `https://database.windows.net/` (which makes sense to me since we are trying to authenticate against a DB, but might be wrong?). We extracted the token from the [`creds.getToken`](https://github.com/Azure/ms-rest-nodeauth/blob/adea8415de97a8076006fbd9948123d09dfdd78d/lib/login.ts#L176) call inside `withUsernamePasswordWithAuthResponse`. The response from `creds.getToken` also contains our tenant id (`tenantId`). However, the [next line](https://github.com/Azure/ms-rest-nodeauth/blob/adea8415de97a8076006fbd9948123d09dfdd78d/lib/login.ts#L179) (`const tenantList = await buildTenantList(creds);`) is the one that fails with the above mentioned errors. If we try to manually modify these two lines (inside `withUsernamePasswordWithAuthResponse` and grab the tenant id and assign it to `tenantList`, we have no problem connecting to the DB and executing queries from our node app:
```javascript
// ...
  const tokenResponse = await creds.getToken();

  // The token cache gets propulated for all the tenants as a part of building the tenantList.
  const tenantList = [ tokenResponse.tenantId ];
// ...
```

so our question(s) are:
1. is there a need to build the tenant list in all situations or can that be grabbed from the response depending on the authorization request (e.g. DB auth) and skip a call to `https://management.azure.com/tenants?api-version=2016-06-01` or is this the intended behavior?
1. if the answer to the above is `we must always make that call and build the list`, is it possible to add `https://database.windows.net/` to the list of allowed audiences? if not, what can we do to successfully authenticate against our Azure SQL DB using Azure AD? (quite possible we are missing a step somewhere :))
1. is `https://database.windows.net/` the right audience? if not, is this something we can modify on our end in Azure (e.g. Azure AD setup, etc.)?

Thank you for looking into our inquiry!

**Comment by amarzavery** [comment]
Yes this is indeed a bug. To keep things simple, if an error occurs while fetching tenants, the library should skip that step and simply move ahead.

To circumvent this: You can either provide a tenantId using the optional parameter `"domain": "<tenant-id>"` or

Install [Azure-CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli-windows?view=azure-cli-latest)
Login via azure cli `az login -h`
and then use
```js
const creds = await AzureCliCredentials.create({resource: "your-resource-id or url"})
```
This should mostly work.

**Comment by daviwil** [comment]
@mir4ef Circling back, are you still experiencing this issue?

**Comment by mir4ef** [comment]
im no longer on this project, but hoping @spalt would be able to provide some update

**Comment by spalt** [comment]
@mir4ef @daviwil still haven't switched back to AD accounts but it's on the schedule for this quarter, so we will know soon if this approach works.  thanks!

**Comment by chintanr97** [comment]
@daviwil and @amarzavery I am facing a same issue: `TypeError: tenants.value is not iterable` while performing interactive login to get a token where the audience is my own AD App. 

`await interactiveLogin({ tokenAudience: ADAppId });`

If I pass `domain: tenantId`, then this issue goes away. However, is there any other way to fix this (without `az login`) for custom resources/audiences?

**Comment by ramya-rao-a** [comment]
@daviwil, I believe the way forward here can be:
- Update `buildTenantList()` to not attempt to loop over `tenants.value` in the case of error response and return an empty array instead
- Update `withUsernamePasswordWithAuthResponse()` to pass the tenantId from the token response to `_getSubscriptions()` if `buildTenantList()` returns an empty array

**Comment by daviwil** [comment]
The fix for this has been shipped in both `2.0.6` and `3.0.6`.  Let us know if you run into any further issues with it!
